File: solutions/06/main.go
package main

import (
	"bufio"
	"errors"
	"flag"
	"log/slog"
	"os"
	"time"

	hashset "github.com/hmcalister/Go-DSA/set/HashSet"
)

const (
	EMPTY_RUNE       rune = '.'
	OBSTACLE_RUNE    rune = '#'
	GUARD_UP_RUNE    rune = '^'
	GUARD_RIGHT_RUNE rune = '>'
	GUARD_DOWN_RUNE  rune = 'v'
	GUARD_LEFT_RUNE  rune = '<'
)

func main() {
	debugFlag := flag.Bool("debug", false, "Debug Flag")
	inputFilePath := flag.String("inputFile", "puzzleInput", "Path to input file.")
	selectedPart := flag.Int("part", 0, "Part to execute. Must be 1 or 2.")
	flag.Parse()
	logFileHandler := SetLogging(*debugFlag)
	defer logFileHandler.Close()

	inputFile, err := os.Open(*inputFilePath)
	if err != nil {
		slog.Error("error opening input file", "error", err)
		os.Exit(1)
	}
	defer inputFile.Close()

	fileScanner := bufio.NewScanner(inputFile)
	if err != nil {
		slog.Error("error processing file input", "error", err)
		os.Exit(1)
	}

	var result int
	computationStartTime := time.Now()
	switch *selectedPart {
	case 1:
		result, err = Part01(fileScanner)
	case 2:
		result, err = Part02(fileScanner)
	default:
		slog.Error("invalid part selected, part must be one of 1 or 2", "part selected", *selectedPart)
		os.Exit(1)
	}
	computationEndTime := time.Now()
	if err != nil {
		slog.Error("error encountered during computation", "error", err, "part selected", *selectedPart)
		os.Exit(1)
	}

	slog.Info("computation completed", "result", result, "computation time elapsed (ns)", computationEndTime.Sub(computationStartTime).Nanoseconds())
}

type MapData struct {
	Width       int
	Height      int
	ObstacleMap *hashset.HashSet[Coordinate]
}

// returns the width, height of the grid, a map of coordinates to obstacles, the guard position and the guard direction
func parseInput(inputLines []string) (MapData, GuardState) {
	guardState := GuardState{
		Coordinate: Coordinate{-1, -1},
		Direction:  DIRECTION_UP,
	}

	obstacleMap := hashset.New[Coordinate]()
	for y, line := range inputLines {
		slog.Debug("read line", "line", line)
		for x, repRune := range line {
			c := Coordinate{x, y}
			switch repRune {
			case OBSTACLE_RUNE:
				slog.Debug("found obstacle", "coordinate", c)
				obstacleMap.Add(c)
			case GUARD_UP_RUNE:
				slog.Debug("found guard up", "coordinate", c)
				guardState.Coordinate = c
				guardState.Direction = DIRECTION_UP
			case GUARD_RIGHT_RUNE:
				slog.Debug("found guard right", "coordinate", c)
				guardState.Coordinate = c
				guardState.Direction = DIRECTION_RIGHT
			case GUARD_DOWN_RUNE:
				slog.Debug("found guard down", "coordinate", c)
				guardState.Coordinate = c
				guardState.Direction = DIRECTION_DOWN
			case GUARD_LEFT_RUNE:
				slog.Debug("found guard left", "coordinate", c)
				guardState.Coordinate = c
				guardState.Direction = DIRECTION_LEFT
			case EMPTY_RUNE:
				// slog.Debug("empty coordinate")
			default:
				slog.Error("unexpected rune found", "rune", repRune, "location x", x, "location y", y)
			}
		}
	}

	if guardState.Coordinate.X == -1 && guardState.Coordinate.Y == -1 {
		slog.Error("no guard found")
		os.Exit(1)
	}

	return MapData{len(inputLines[0]), len(inputLines), obstacleMap}, guardState
}

// Count the number of visited cells (not states, direction is irrelevant) in path set by the initial guard state
// If the path loops at any point, an error is returned
func (m MapData) CheckVisitedCells(guardState GuardState) (int, error) {
	visitedCellsSet := hashset.New[Coordinate]()
	visitedStatesSet := hashset.New[GuardState]()

	for guardState.InBounds(m.Width, m.Height) {
		if visitedStatesSet.Contains(guardState) {
			// We have seen this state before, therefore we are in a loop
			return -1, errors.New("a loop has occurred in the path")
		}
		visitedStatesSet.Add(guardState)
		visitedCellsSet.Add(guardState.Coordinate)

		nextState := guardState.Step()
		if m.ObstacleMap.Contains(nextState.Coordinate) {
			// slog.Debug("found obstacle", "current state", guardState, "numVisitedCells", len(visitedCells))
			guardState = guardState.EncounterObstacle()
			continue
		}

		// slog.Debug("making step", "current state", guardState, "numVisitedCells", len(visitedCells))
		guardState = nextState
	}

	return visitedCellsSet.Size(), nil
}

func Part01(fileScanner *bufio.Scanner) (int, error) {
	inputLines := make([]string, 0)
	for fileScanner.Scan() {
		inputLines = append(inputLines, fileScanner.Text())
	}
	mapData, guardState := parseInput(inputLines)
	slog.Debug("parsed input", "mapWidth", mapData.Width, "mapHeight", mapData.Height, "obstacleMap", mapData.ObstacleMap, "guardState", guardState)

	numVisitedCells, err := mapData.CheckVisitedCells(guardState)
	if err != nil {
		slog.Error("error occurred when checking path", "error", err)
		os.Exit(1)
	}

	return numVisitedCells, nil
}

func Part02(fileScanner *bufio.Scanner) (int, error) {
	inputLines := make([]string, 0)
	for fileScanner.Scan() {
		inputLines = append(inputLines, fileScanner.Text())
	}
	mapData, guardState := parseInput(inputLines)
	slog.Debug("parsed input", "mapWidth", mapData.Width, "mapHeight", mapData.Height, "obstacleMap", mapData.ObstacleMap, "guardState", guardState)

	visitedStatesSet := hashset.New[Coordinate]()
	loopCreatingObstacleSet := hashset.New[Coordinate]()

	// Walk over the path and at each step (that does not already have an obstacle in front of it)
	// see if adding an obstacle introduces a loop. If so, count it. Otherwise, remove the obstacle and take a step.
	for guardState.InBounds(mapData.Width, mapData.Height) {
		visitedStatesSet.Add(guardState.Coordinate)

		nextState := guardState.Step()
		if mapData.ObstacleMap.Contains(nextState.Coordinate) {
			slog.Debug("found existing obstacle", "current state", guardState)
			guardState = guardState.EncounterObstacle()
			continue
		}

		// Only add an obstacle if the next state is in bounds and the coordinate has not been visited already (blocking the previous path...)
		if !visitedStatesSet.Contains(nextState.Coordinate) && nextState.InBounds(mapData.Width, mapData.Height) {
			slog.Debug("try path with obstacle inserted", "current state", guardState)
			mapData.ObstacleMap.Add(nextState.Coordinate)
			_, err := mapData.CheckVisitedCells(guardState)
			if err != nil {
				slog.Debug("loop encountered with obstacle", "obstacle coordinate", nextState.Coordinate)
				loopCreatingObstacleSet.Add(nextState.Coordinate)
			}
			mapData.ObstacleMap.Remove(nextState.Coordinate)
		}

		slog.Debug("making step", "current state", guardState)
		guardState = nextState
	}

	return loopCreatingObstacleSet.Size(), nil
}

File: solutions/06/src/main.rs
use std::collections::{HashMap, HashSet};
use std::fs::File;
use std::io::{BufRead, BufReader};

use clap::Parser;
use std::time::SystemTime;
#[allow(unused_imports)]
use tracing::{debug, error, info, trace};
use tracing::{span, Level};

mod guard_state;
mod logging;
mod obstacle_map;

use guard_state::coordinate;
use guard_state::coordinate::Coordinate;
use guard_state::direction::Direction;
use guard_state::GuardState;
use obstacle_map::ObstacleMap;

const GUARD_INITIAL_DIRECTION_UP: u8 = b'^';
const GUARD_INITIAL_DIRECTION_RIGHT: u8 = b'>';
const GUARD_INITIAL_DIRECTION_DOWN: u8 = b'v';
const GUARD_INITIAL_DIRECTION_LEFT: u8 = b'<';

/// Program to solve Advent of Code puzzles
#[derive(Parser, Debug)]
#[command(version, about)]
struct CommandLineArgs {
    /// File to read input from
    #[arg(long, default_value = "puzzleInput")]
    input_file: String,

    /// Selected part of puzzle to solve
    #[arg(long, value_parser=clap::value_parser!(u8).range(1..=2))]
    part: u8,
}

fn main() {
    logging::set_logging();
    let args = CommandLineArgs::parse();
    trace!("command line args" = ?args, "parsed command line args");

    let input_file_handle = File::open(args.input_file).expect("could not open input file");
    let input_file_reader = BufReader::new(input_file_handle);

    let start_time = SystemTime::now();
    let computation_result = match args.part {
        1 => part01(input_file_reader),
        2 => part02(input_file_reader),
        _ => unreachable!(), // clap has filtered out all other possibilities.
    }
    .expect("computation did not produce a value");
    let elapsed_time = start_time.elapsed().unwrap();

    info!(
        "computation_result" = computation_result,
        "elapsed_time" = elapsed_time.as_nanos(),
        "computation complete"
    );
}

fn parse_input_to_obstacle_map_and_state(
    input_file_reader: BufReader<File>,
) -> (ObstacleMap, Option<GuardState>) {
    
    let all_lines: Vec<_> = input_file_reader
    .lines()
    .collect::<Result<Vec<_>, _>>()
    .unwrap();
    let mut obs_map = obstacle_map::new(all_lines[0].len() as i32, all_lines[0].len() as i32);
    let mut guard: Option<GuardState> = None;

    for (y, line) in all_lines.iter().enumerate() {
        debug!("line" = ?line, "read line from input file");
        for (x, r) in line.bytes().enumerate() {
            let c = coordinate::new_from_usize(x, y);
            let _span = span!(Level::DEBUG, "parsed byte", "coordinate"=?c, "byte"= r).entered();
            match r {
                obstacle_map::OBSTACLE_BYTE => {
                    trace!("found obstacle");
                    obs_map.add_obstacle(c);
                }
                GUARD_INITIAL_DIRECTION_UP => {
                    trace!("found guard facing up");
                    guard = Some(guard_state::new(c, Direction::Up));
                }
                GUARD_INITIAL_DIRECTION_RIGHT => {
                    trace!("found guard facing right");
                    guard = Some(guard_state::new(c, Direction::Right));
                }
                GUARD_INITIAL_DIRECTION_DOWN => {
                    trace!("found guard facing down");
                    guard = Some(guard_state::new(c, Direction::Down));
                }
                GUARD_INITIAL_DIRECTION_LEFT => {
                    trace!("found guard facing left");
                    guard = Some(guard_state::new(c, Direction::Left));
                }
                _ => {
                    trace!("found other byte");
                }
            }
        }
    }

    (obs_map, guard)
}

/// Compute the number of steps required to leave the bounds of the map.
/// 
/// Returns None if the guard loops.
fn compute_path(obs_map: &ObstacleMap, initial_state: GuardState) -> Option<Vec<Coordinate>> {
    let _span = span!(Level::DEBUG, "computing path", "initial_state"=?initial_state,).entered();
    let mut seen_states: HashSet<GuardState> = HashSet::new();
    let mut seen_coordinates: Vec<Coordinate> = Vec::new();
    let mut current_state = initial_state;
    let mut total_steps = 0;

    loop {
        if seen_states.contains(&current_state) {
            return None
        }
        seen_states.insert(current_state);
        seen_coordinates.push(current_state.get_coordinate());
        
        let next_state = current_state.step();
        match obs_map.is_obstacle(next_state.get_coordinate()) {
            Some(false) => {
                total_steps += 1;
                current_state = next_state;
                debug!("current state"=?current_state, "total steps"=total_steps, "took step");
            },
            Some(true) => {
                debug!("current state"=?current_state, "total steps"=total_steps, "encountered obstacle");
                current_state = current_state.encounter_obstacle();
            },
            None => {
                debug!("current state"=?current_state,"left map");
                return Some(seen_coordinates);
            }
        }
    }
}

fn part01(input_file_reader: BufReader<File>) -> Option<i64> {
    let (obs_map, guard) = match parse_input_to_obstacle_map_and_state(input_file_reader) {
        (obs_map, Some(guard)) => (obs_map, guard),
        (_, None) => {
            error!("failed to initialize guard state from input");
            return None;
        }
    };
    debug!("obstacle map"=?obs_map, "initial state"=?guard, "parsed input");

    match compute_path(&obs_map, guard) {
        Some(seen_coordinates) => {
            let unique_coordinates = seen_coordinates.into_iter().collect::<HashSet<_>>();
            return Some(unique_coordinates.len() as i64)
        },
        None => {
            error!("did not find a vlid path for part 01");
            return None;
        }
    }

}

fn part02(input_file_reader: BufReader<File>) -> Option<i64> {
}
